src/crimson/osd/osd_operation_sequencer.h

```diff
@@ -0,0 +1,102 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <map>
+
+#include "crimson/common/operation.h"
+#include "osd/osd_types.h"
+
+namespace crimson::osd {
+
+template <typename>
+struct OperationComparator;
+
+template <typename T>
+class OperationRepeatSequencer {
+public:
+  using OpRef = boost::intrusive_ptr<T>;
+  using ops_sequence_t = std::map<OpRef, seastar::promise<>>;
+
+  template <typename Func, typename HandleT, typename Result = std::invoke_result_t<Func>>
+  seastar::futurize_t<Result> start_op(
+      HandleT& handle,
+      epoch_t same_interval_since,
+      OpRef& op,
+      const spg_t& pgid,
+      Func&& func) {
+    auto& ops = pg_ops[pgid];
+    if (!op->pos) {
+      auto [it, inserted] = ops.emplace(op, seastar::promise<>());
+      assert(inserted);
+      op->pos = it;
+    }
+
+    const auto p = *(op->pos);
+    const bool first = (p == ops.begin());
+
+    this->same_interval_since = same_interval_since;
+
+    typename OperationRepeatSequencer<T>::ops_sequence_t::iterator last_op_it = *(op->pos);
```

Comment:
no need to define `last_op_it`, we already have `p`. probably it deserves a better name though.